### PR TITLE
Implements API schema spec

### PR DESCRIPTION
**Pull request by AmineAfia**
### TL;DR

Implemented new API endpoints for transactions and events, enhanced query parameter handling, and improved error responses.

### What changed?

- Added new API endpoints for transactions and events, including chain-specific, contract-specific, and function signature-specific routes.
- Implemented `ParseQueryParams` function to handle complex query parameters, including filtering, grouping, sorting, and pagination.
- Enhanced error handling with more specific error types (BadRequest, Unauthorized, Internal).
- Updated the `QueryResponse` and `Meta` structs to include more detailed information about the query results.
- Added a health check endpoint.
- Removed the `GetBlocks` handler and replaced it with more specific transaction and event handlers.
- Updated dependencies, including upgrading chi and gorilla/schema.

### How to test?

1. Run the API server locally.
2. Test the new endpoints:
   - GET /api/transactions/{chainId}
   - GET /api/events/{chainId}
   - GET /api/transactions/{chainId}/{contractAddress}
   - GET /api/events/{chainId}/{contractAddress}
   - GET /api/transactions/{chainId}/{contractAddress}/{functionSig}
   - GET /api/events/{chainId}/{contractAddress}/{functionSig}
3. Test the health check endpoint: GET /health
4. Verify that query parameters are correctly parsed and included in the response.
5. Test error scenarios, including unauthorized access and invalid chain IDs.

### Why make this change?

This change aims to provide a more flexible and powerful API for querying blockchain data. The new endpoints allow for more granular queries, while the enhanced query parameter handling supports advanced filtering, sorting, and aggregation. The improved error handling and response structure will make it easier for clients to consume and understand the API responses.

### CODE REVIEW

**Comment by AmineAfia** [comment]
* **#39** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/39?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @AmineAfia and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/handlers/api.go [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
+		router.Get("/transactions/{chainId}", GetTransactions)
+		router.Get("/events/{chainId}", GetLogs)
+
+		// contract scoped queries
+		router.Get("/transactions/{chainId}/{contractAddress}", GetTransactionsWithContract)
+		router.Get("/events/{chainId}/{contractAddress}", GetEventsWithContract)
+
+		// signature scoped queries
+		router.Get("/transactions/{chainId}/{contractAddress}/{functionSig}", GetTransactionsWithContractAndSignature)
+		router.Get("/events/{chainId}/{contractAddress}/{functionSig}", GetEventsWithContractAndSignature)
```
Comment (iuwqyir) [inline internal/handlers/api.go:26]: Doesn't it make more sense from a REST viewpoint to have `chainId` as the first resource?
e.g. `api/{chainId}/transactions`

File: api/api.go
Comment (iuwqyir) [inline api/api.go:1]: Why is this under `/api`? Based on the layout doc, this folder is reserved for `OpenAPI/Swagger specs, JSON schema files, protocol definition files.`

File: api/api.go [hunk 2]
```diff
@@ -3,26 +3,52 @@ package api
...
-	Code int
-	Message string
+	Code      int    `json:"code"`
+	Message   string `json:"message"`
+	SupportId string `json:"support_id"`
 }
 
 type QueryParams struct {
-	ChainID string
+	FilterParams map[string]string `schema:"-"`
```
Comment (iuwqyir) [inline api/api.go:22]: Can you do larger/smaller/equal here?

Comment (AmineAfia) [reply api/api.go:22]: yes according to the spec filter operators are bart of the filter key. e.g. `filter_block_number_gte=6666`

File: api/api.go
Comment (AmineAfia) [reply api/api.go:1]: You mean the functions? if so the functions are validators for the schema

File: internal/handlers/api.go (hunk 1 above)
Comment (AmineAfia) [reply internal/handlers/api.go]: done

**Review by iuwqyir — APPROVED** [review]

**Pull request merged**